cs4660/graph/graph.py

```diff
@@ -0,0 +1,544 @@
+        
+            for nodes, edges in self.adjacency_list.items():
+                temp_list = []
+                for edge in edges:
+                    if edge.to_node != node:
+                        temp_list.append(edge)
+                self.adjacency_list[nodes] = temp_list
+            return True
+        else: return False
+
+        """
+                if node in self.adjacency_list:
+                    self.adjacency_list.remove(node)
+            del self.adjacency_list[node]
+            return True 
+        """
+
+        # add edge complete
+    def add_edge(self, edge):
+        """if edge.from_node not in self.adjacency_list or edge.to_node not in self.adjacency_list:
+            return False
+
+        my_edges = self.adjacency_list[edge.from_node]
+        for my_edge in my_edges:
+            if edge == my_edge:
+                return False
+        
+        my_edges.append(edge)
+        return True
+        """
+        neighbors = self.adjacency_list[edge.from_node]
+        if edge not in neighbors:
+            neighbors.append(edge)
+            return True
+        else: return False 
+
+
+
+        # remove edge complete
+    def remove_edge(self, edge):
+        """self.edge.adjacency_list.discard(edge)"""
+        
+        # which is better adjacency list or adjacency matrix: https://stackoverflow.com/questions/2218322/what-is-better-adjacency-lists-or-adjacency-matrices-for-graph-problems-in-c
+        """if edge.from_node in self.adjacency_list:
+            return False
+        if edge.to_node in self.adjacency_list[edge.from_node]:
+            self.adjacency_list[edge.from_node].remove(edge.to_node)
+            return True
+        """
+        # cleaner this way
+        """adj_edge_to_remove = self.adjacency_list[edge.from_node]
+        if edge not in adj_edge_to_remove:
+            return False
+        adj_edge_to_remove.remove(edge)
+        return True"""
+        neighbors = self.adjacency_list[edge.from_node]
+        if edge in neighbors:
+            neighbors.remove(edge)
+            return True
+        else: return False 
+
+# We have triple check this part of the code for debugging purposes.
+
+class AdjacencyMatrix(object):
+    def __init__(self):
+        # adjacency_matrix should be a two dimensions array of numbers that
+        # represents how one node connects to another
+        self.adjacency_matrix = []
+        # in additional to the matrix, you will also need to store a list of Nodes
+        # as separate list of nodes
+        self.nodes = []
+        # complete
+    def adjacent(self, node_1, node_2):
+        """if node_1 not in self.nodes or node_2 not in self.nodes:
+            return False
+        if self.adjacency_matrix[node_1.data][node_2.data] != 0:
+            return True
+        """
+        for edge in self.adjacency_matrix[self.__get_node_index(node_1)]:
+            if edge.to_node == node_2:
+                return True
```

Comment:
data structure used here for the adjacency matrix is wrong.

Reply:
The data adjacency_matrix you code here is a dictionary of node to a list of edges, which is *a lot* different than two dimensional array of numbers.

Reply:
Thanks a lot, prof.